src/librustc/ty/contents.rs

```diff
@@ -235,6 +235,11 @@ impl<'a, 'tcx> ty::TyS<'tcx> {
                             })
                         });
 
+                    if def.is_union() {
+                        // unions don't have destructors regardless of the child types
```

Comment:
Could you add a test checking that `needs_drop` still returns `true` on unions that implement `Drop` themselves?
I suspect the condition should be `def.is_union() && !def.has_dtor()` instead of just `def.is_union()`.